Add RNE Derivative Support

zer-art opened the pull request:
This PR implements the implicit velocity integrator for `mujoco_warp`, mirroring the logic found in the C MuJoCo engine. This includes the implementation of the Recursive Newton-Euler (RNE) derivative (`mjd_rne_vel`) which is required for the implicit integrator's Jacobian computation.

This addresses the requirements discussed in #891.

### Key Changes

* **`mujoco_warp/_src/derivative.py`**:
    * Implemented `mjd_rne_vel` as the main entry point for derivative computation.
    * Added forward/backward pass kernels and sparse/dense update kernels for `qDeriv`.
* **`mujoco_warp/_src/forward.py`**:
    * Updated `implicit` to handle `IntegratorType.IMPLICIT` and `IMPLICITFAST`.
    * Integrated `smooth.factor_solve_i` (LDL factorization) as the linear solver.
* **`mujoco_warp/_src/io.py`**:
    * Fixed strict type checking issues for `wp.full` inputs.
* **Tests**:
    * Updated `forward_test.py` to include `IntegratorType.IMPLICIT`.

### Verification
- [x] Code compiles with Warp 1.10.1
- [x] Implemented `mjd_rne_vel` logic
- [x] Updated `implicit` integrator handling
- [x] Added test cases in `forward_test.py`

### Related Issue
Closes #891

zer-art committed:
Add Implicit Integrator and RNE Derivative Support

thowell commented:
@zer-art thank you for contributing this pr!

let's focus this pr on the warp implementation of `mjd_rne_vel`: `deriv_rne_vel` and test the implementation in `derivative_test.py`. we should also integrate this function in `deriv_smooth_vel` and test this as well. 

we can complete the implicit integrator feature #891 with additional prs. after this pr, it probably makes sense to follow up with:
- LU solver pr based on the mujoco implementation
- implicit integrator pr that integrates and tests everything

please comment if you have any questions. thanks!

zer-art committed:
refactor: scope PR to RNE derivative implementation only

zer-art committed:
Merge branch 'main' into Implicit-Integrator-and-RNE-Derivative-Support

zer-art committed:
fixed linting issues

thowell commented:
@zer-art

3. please add the test to derivative_test.py

thanks!

zer-art committed:
Merge branch 'google-deepmind:main' into Implicit-Integrator-and-RNE-Derivative-Support

zer-art committed:
Move RNE-specific derivative tests into `derivative_test.py` .

zer-art committed:
Merge branch 'main' into Implicit-Integrator-and-RNE-Derivative-Support

zer-art committed:
Merge branch 'google-deepmind:main' into Implicit-Integrator-and-RNE-Derivative-Support

zer-art commented on mujoco_warp/_src/derivative.py:
```diff
@@ -288,4 +289,366 @@ def deriv_smooth_vel(m: Model, d: Data, out: wp.array2d(dtype=float)):
+      elif k >= dofid + 3 and k < dofid + 6:
+        cvel_k += cdof[k]
+
+      # Dcdofdot = cross(Dcvel, cdof)
```
Done. I've cleaned up all the verbose comments and implementation notes in the file.

zer-art committed:
feat: implement analytical RNE derivatives and consolidate tests

- derivative.py: Implemented RNE Term 2 using `motion_cross_force` to correctly calculate momentum derivatives. Cleaned up comments and TODOs.
- derivative_test.py: Merged RNE stress tests into `test_rne_stress`, replacing `test_rne_vel_effect`.
- types.py: Marked `IMPLICIT` integrator as unsupported.

zer-art commented on mujoco_warp/_src/derivative.py:
```diff
@@ -288,4 +289,366 @@ def deriv_smooth_vel(m: Model, d: Data, out: wp.array2d(dtype=float)):
+  # Dcacc accumulation
+  # ...
+  # Placeholder for complex logic, I'll complete this in next step.
+  pass
```
I decided to fully implement the RNE logic (rne_vel) in this PR instead of leaving a TODO. The forward and backward passes are now implemented in `derivative.py` and verified by `test_rne_stress`.

zer-art committed:
fix comments

zer-art committed:
fix linting issues

zer-art committed:
Merge branch 'google-deepmind:main' into Implicit-Integrator-and-RNE-Derivative-Support

zer-art committed:
Merge branch 'google-deepmind:main' into Implicit-Integrator-and-RNE-Derivative-Support

zer-art committed:
feat: resolve conflicts and fix MuJoCo 3.5.0 compatibility

zer-art committed:
Merge branch 'main' into Implicit-Integrator-and-RNE-Derivative-Support

thowell committed:
Merge upstream/main into PR #912

thowell committed:
Merge upstream/main into PR #912

thowell committed:
update rne derivatives implementation

thowell commented:
@zer-art the pr branch is updated. please have a look at the changes. thanks!

zer-art commented:
Hi @thowell, I've pulled your latest changes and ran the full suite in derivative_test.py on my local system. All 18 tests (including the parameterized checks for dense and sparse) are passing successfully. The implementation looks solid and correctly resolves the previous issues. Ready to merge from my side! Thanks for the help.

thowell committed:
Merge remote-tracking branch 'upstream/main' into pr912-merge

thowell committed:
updates

btaba commented on mujoco_warp/_src/derivative_test.py:
```diff
@@ -534,6 +539,250 @@ def test_forcerange_clamped_derivative(self):
+      </worldbody>
+      <keyframe>
+        <key qpos="0.5 0.02 1 0 0 0 0.7"
+             qvel="2.0 -0.5 0.3 0.8 -1.2 1.0"/>
```
Looking at MuJoCo tests, tumbling_thin_body might be a good addition to these tests. Other than that, the test coverage looks OK,  but I'm curious how you picked these qvel?

thowell replied:
not sure about the previous keyframe selection.

tests are updated to include additional scenes from mujoco tests and the keyframes are now based on simulating the scene for 100 timesteps.

thowell committed:
update derivative_test.py
